### ISSUE TITLE
size mismatch for module.prediction.weight/bias

### ISSUE THREAD

**Issue by AtiBabaie** [ku21fan/STR-Fewer-Labels#15]
Hi,

I have done the pre-train and train part with no errors and got pretty good accuracy. but for the test part I have got this error that says the torch size is different of 216 and 218.  you can see the full error below.
I would appreciate any help.
Thanks in advance.

test error:

![4_test_3](https://github.com/ku21fan/STR-Fewer-Labels/assets/136967442/fa5ad946-745f-4dc7-80ae-90b9d30cc4d6)
![4_test_2](https://github.com/ku21fan/STR-Fewer-Labels/assets/136967442/bb38db93-7b74-4208-9e0e-446bde98fbf5)

train accuracy:

![4_train_23](https://github.com/ku21fan/STR-Fewer-Labels/assets/136967442/a50921c8-2d48-4326-b704-6a40e1dc1ae9)

**Comment by ku21fan** [comment]
Hi, 

Glad to hear that your training went well!

This error indicates a discrepancy in the number of character classes. 
From your error message, your trained model used 216 for character classes, 
but in your test code, `# of tokens and characters: 218`
From this, I guess your test code uses 2 additional characters compared to the training code. 

Please check the character set of your train.py and test.py

Hope it helps, 
Jeonghun

**Comment by AtiBabaie** [comment]
Thank you Jeonghun,

Yes, I've realized that, the character sets were different and I fixed it and got the test accuracy too.  :)
but I have another question. There are 2 characters that exist in the character set but it shows unknown in the output.
As you can see in train accuracy image up there.
The 2 characters are : آ and ئـ
I think the net consider each as 2 separated character. cause the output is : ا + ~ on the top of it.
And  same for the latter one.
here is another example :

![4_train_18](https://github.com/ku21fan/STR-Fewer-Labels/assets/136967442/c915b9aa-3499-42cd-bef1-1b1c555a213e)

Do you have any idea on this case?
Thanks again.

**Comment by ku21fan** [comment]
Hi Ati, 

I am not familiar with the Arabic language, so I am not sure what is an effective solution.
I guess the below part separates آ into two characters.
https://github.com/ku21fan/STR-Fewer-Labels/blob/e6aa817e2eacbf29b3fcd11390d78b1a8f96bf78/utils.py#L52

In my opinion, there is a workaround using substitute characters for آ and ئـ as shown in the below issue (2)
https://github.com/clovaai/deep-text-recognition-benchmark/issues/85#issuecomment-592019386

Hope it helps,
Jeonghun

**Comment by AtiBabaie** [comment]
Yes, thank you so much.
:)

**Comment by mujeeb-merwat** [comment]
@AtiBabaie I'm working on Urdu OCR, I tried a different way, changed the
character to Urdu: ا ب پ ت ٹ ث ج چ ح خ د ڈ ذ ر ڑ ز ژ س ش ص ض ط ظ ع غ ف ق ک گ ل م ن ں و ہ ھ ء ی ے, but it didn't work. When I'm loading pre trained model "TPS-ResNet-BiLSTM-Attn.pth", it gives dimensions error with number of classes.

My question is which model you have fintuned for Arabic?  And you did the same way by adding charecter list simply.? Would love your help.

Thanks

**Comment by AtefeBabaie** [comment]
> @AtiBabaie I'm working on Urdu OCR, I tried a different way, changed the character to Urdu: ا ب پ ت ٹ ث ج چ ح خ د ڈ ذ ر ڑ ز ژ س ش ص ض ط ظ ع غ ف ق ک گ ل م ن ں و ہ ھ ء ی ے, but it didn't work. When I'm loading pre trained model "TPS-ResNet-BiLSTM-Attn.pth", it gives dimensions error with number of classes.
> 
> My question is which model you have fintuned for Arabic? And you did the same way by adding charecter list simply.? Would love your help.
> 
> Thanks

Hi mujeeb,

My language is Persian which arabic and Urdu are close to it so you can do the same.
I used **"Pretrain with RotNet (For TRBA, use --model_name NR)"** for pretraining and **"Train CRNN with RotNet initialization"** for training.
For your second question, yes! I simply put all the characters in the list.
